object_store/src/throttle.rs

```diff
@@ -0,0 +1,500 @@
+
+        assert_bounds!(measure_list_with_delimiter(&store, 0).await, 0);
+        assert_bounds!(measure_list_with_delimiter(&store, 10).await, 0);
+
+        store.wait_list_with_delimiter_per_call = WAIT_TIME;
+        assert_bounds!(measure_list_with_delimiter(&store, 0).await, 1);
+        assert_bounds!(measure_list_with_delimiter(&store, 10).await, 1);
+
+        store.wait_list_with_delimiter_per_call = ZERO;
+        store.wait_list_with_delimiter_per_entry = WAIT_TIME;
+        assert_bounds!(measure_list_with_delimiter(&store, 2).await, 2);
+
+        store.wait_list_with_delimiter_per_call = WAIT_TIME;
+        store.wait_list_with_delimiter_per_entry = WAIT_TIME;
+        assert_bounds!(measure_list_with_delimiter(&store, 2).await, 3);
+    }
+
+    #[tokio::test]
+    async fn put_test() {
+        let inner = InMemory::new();
+        let mut store = ThrottledStore::new(inner);
+
+        assert_bounds!(measure_put(&store, 0).await, 0);
+        assert_bounds!(measure_put(&store, 10).await, 0);
+
+        store.wait_put_per_call = WAIT_TIME;
+        assert_bounds!(measure_put(&store, 0).await, 1);
+        assert_bounds!(measure_put(&store, 10).await, 1);
+
+        store.wait_put_per_call = ZERO;
+        store.wait_put_per_byte = WAIT_TIME;
+        assert_bounds!(measure_put(&store, 2).await, 2);
+
+        store.wait_put_per_call = WAIT_TIME;
+        store.wait_put_per_byte = WAIT_TIME;
+        assert_bounds!(measure_put(&store, 2).await, 3);
+    }
+
+    async fn place_test_object(
+        store: &ThrottledStore<InMemory>,
+        n_bytes: Option<usize>,
+    ) -> <ThrottledStore<InMemory> as ObjectStoreApi>::Path {
+        let mut path = store.new_path();
+        path.set_file_name("foo");
+
+        if let Some(n_bytes) = n_bytes {
+            let data = Bytes::from_iter((0..n_bytes).map(|_| 1u8));
+            let stream_data = std::io::Result::Ok(data.clone());
+            let stream = futures::stream::once(async move { stream_data });
+            store.put(&path, stream, None).await.unwrap();
+        } else {
+            // ensure object is absent
+            store.delete(&path).await.unwrap();
+        }
+
+        path
+    }
+
+    async fn place_test_objects(
+        store: &ThrottledStore<InMemory>,
+        n_entries: usize,
+    ) -> <ThrottledStore<InMemory> as ObjectStoreApi>::Path {
+        let mut prefix = store.new_path();
+        prefix.push_dir("foo");
+
+        // clean up store
+        for entry in store
+            .list(Some(&prefix))
+            .await
+            .unwrap()
+            .try_concat()
+            .await
+            .unwrap()
+        {
+            store.delete(&entry).await.unwrap();
+        }
+
+        // create new entries
+        for i in 0..n_entries {
+            let mut path = prefix.clone();
+            path.set_file_name(&i.to_string());
+
+            let data = Bytes::from("bar");
+            let stream_data = std::io::Result::Ok(data.clone());
```

Comment:
Same with this clone:

```suggestion
            let stream_data = std::io::Result::Ok(data);
```